cms/djangoapps/contentstore/tasks.py

```diff
@@ -1066,3 +1070,186 @@ def undo_all_library_source_blocks_ids_for_course(course_key_string, v1_to_v2_li
+        """
+        key = arguments_dict['course_key_string']
+        return f'Broken link check of {key}'
+
+
+@shared_task(base=CourseLinkCheckTask, bind=True)
+def check_broken_links(self, user_id, course_key_string, language):
+    """
+    Checks for broken links in a course. Store the results in a file.
+    """
+    courselike_key = CourseKey.from_string(course_key_string)
+
+    try:
+        user = User.objects.get(pk=user_id)
+    except User.DoesNotExist:
+        with translation_language(language):
+            self.status.fail(UserErrors.UNKNOWN_USER_ID.format(user_id))
+        return
+    if not has_course_author_access(user, courselike_key):
+        with translation_language(language):
+            self.status.fail(UserErrors.PERMISSION_DENIED)
+        return
+    
+    courselike_block = modulestore().get_course(courselike_key)
+
+    try:
+        self.status.set_state('Scanning')
+        links_file = _create_broken_link_json(courselike_block, courselike_key, {}, self.status)
+        artifact = UserTaskArtifact(status=self.status, name='BrokenLinks')
+        artifact.file.save(name=os.path.basename(links_file.name), content=File(links_file))
+        artifact.save()
+    # catch all exceptions so we can record useful error messages
+    except Exception as exception:  # pylint: disable=broad-except
+        LOGGER.exception('Error checking links for course %s', courselike_key, exc_info=True)
+        if self.status.state != UserTaskStatus.FAILED:
+            self.status.fail({'raw_error_msg': str(exception)})
+        return
+
+
+def _create_broken_link_json(course_block, course_key, context, status=None):
+    """
+    Generates a json file for broken links, or returns None if there was an error.
+    Updates the context with any error information if applicable.
+    Note that because the celery queue does not have credentials, some broken links will
+    need to be checked client side.
+
+    File format:
+    [
+        [<blockId>, <broken link>],
+        ...,
+    ]
+    """
+    name = course_block.url_name
+    links_file = NamedTemporaryFile(prefix=name + '.', suffix='.json')
+
+    try:
+        if status:
+            status.set_state('Verifying')
+            status.increment_completed_steps()
+        LOGGER.debug('json file being generated at %s', links_file.name)
+
+        verticals = modulestore().get_items(
+            course_key,
+            qualifiers={'category': 'vertical'},
+        )
+        blocks = []
+        for vertical in verticals:
+            blocks.extend(vertical.get_children())
+        data = []
+        for block in blocks:
+            usage_key = block.usage_key
+            block_info = get_block_info(block)
+            block_data = block_info['data']
+            urls = _get_urls(block_data)
+
+            for url in urls:
+                if url == '#':
+                    break
+                processed_url = _process_url(url, course_key)
```

Comment:
I recommend steering away from the word "process" in method names, as in it misses the opportunity to shed light on the work that's being done. In this case, it's recasting the URL into a standardized form, prior to checking whether it's valid or not. So why not name the method `_cast_url_to_standardized_form()`. Or, if I've misunderstood, to something equivalently meaningful?